internal/ratelimit/ratelimit.go

```diff
@@ -0,0 +1,185 @@
+package ratelimit
+
+import (
+	"net/http"
+	"regexp"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+)
+
+// Ratelimiter holds all ratelimit buckets
+type RateLimiter struct {
+	sync.Mutex
+	buckets         map[string]*Bucket
```

Comment:
You mean each bucket having a reference to a global bucket instead of the ratelimiter?